File: 프로그래머스/3/12927. 야근 지수/야근 지수.java
import java.util.*;

class Solution {
    public long solution(int n, int[] works) {
        
        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
        
        for(int i=0; i<works.length; i++){
            pq.add(works[i]);
        }
        
        while(n > 0 && !pq.isEmpty()){
            if(!pq.isEmpty()){
                int a = pq.poll();
                a--;
                n--;
                if(a != 0) pq.add(a);
            }
        }
        
        long answer = 0;
        
        while(!pq.isEmpty()){
            int now = pq.poll();
            answer += now*now;
        }
        
        return answer;
    }
}

File: 프로그래머스/3/12927. 야근 지수/야근 지수.js
function solution(n, works) {
    
    if(works.reduce((a, b) => a+b) <= n) return 0;
    
    var sortWork = works.sort((a,b) => a-b)
    
    while(n > 0 && works.length > 0){
        
        var max = sortWork[works.length-1];
        
        for(let i=works.length-1; i>=0; i--){
            if(sortWork[i] >= max){
                sortWork[i]--;
                n--;
            }
            if(n==0) break;
        }
        
    }
    
    var answer = 0;
    
    while(works.length){
        var now = works.pop();
        answer += now*now;
    }
    
    return answer;
}